[docs] Add a note about testing websites served over a VPN (#5758)
DevExpress/testcafe: docs/articles/faq/README.md

* [TravisCI](../documentation/guides/continuous-integration/travis.md)
* [CircleCI](../documentation/guides/continuous-integration/circleci.md)
* [AppVeyor](../documentation/guides/continuous-integration/appveyor.md)

### 'The browser can't open the page: can't establish a secure connection to the server'

> If your computer is connected to multiple networks (for instance, if it uses a VPN connection), TestCafe may be unable to open the testing pages.

Author: ```suggestion
If your computer is connected to multiple networks (for instance, if it uses a VPN connection), TestCafe may detect the host IP incorrectly and thus may be unable to open the testing pages.
```

> If your computer is connected to multiple networks (for instance, if it uses a VPN connection), TestCafe may detect the host IP incorrectly and thus may be unable to open the testing pages.

Reviewer: ```suggestion
If your computer is connected to multiple networks (for instance, if it uses a VPN connection), TestCafe may incorrectly detect the host IP and not open the tested pages.
```

> If your computer is connected to multiple networks (for instance, uses a VPN connection), TestCafe may host its internal proxy server on an IP that cannot be routed correctly due to VPN. In this case, the following error occurs:

Reviewer 2: `TestCafe may host its internal proxy server on an IP` - it's too complicated.

Can be simplified as `TestCafe can calculate an host IP that cannot be reached/accesses...`

Author: Perhaps

> If your computer is connected to multiple networks (for instance, uses a VPN connection), TestCafe may pick an IP that is inaccessible due to VPN.

But then it's unclear ('pick for what?', 'unaccessible from where?'). 

@VasilyStrelyaev WDYT?

Reviewer 3: Maybe: 

> If your computer is connected to multiple networks (for instance, if it uses VPN), TestCafe may identify the host with an IP that does not resolve to `localhost` due to VPN or routing issues.

Reviewer 2: No.

The problem - By default, TestCafe uses the machine IP as host for the proxied URLs. When you connect via VPN, you IP can be calculated as IP of machine FROM you come. 
So, the solution - prevent using IP and use the 'localhost' alias.

Reviewer 3: Oh, I see, I thought the problem was with reaching the correct IP...

Okay, cool, thanks, so:

> If your computer is connected to multiple networks (for instance, if it uses a VPN connection), TestCafe may detect the host IP incorrectly and thus generate testing links that do not open.

Reviewer 2: > and thus generate testing links that do not open

It's not clear. Which testing links?

What about such variant?
`and cannot open the testing pages`.

Author: ```suggestion
If your computer is connected to multiple networks (for instance, if it uses a VPN connection), TestCafe may be unable to open the testing pages.
```

```text

> <browsername> can't open the page because <browsername> can't establish a secure connection to the server.

Reviewer: ```suggestion
<browsername> cannot open the page because <browsername> is unable to establish a secure connection to the server.
```

```

> To explicitly state that the proxy should be hosted on the local machine, launch TestCafe with the `--hostname localhost` CLI option:

Reviewer 2: `To explicitly state that the proxy should be hosted on the local machine`  - it's not correct.
The 'hostname' parameter just guarantees that the proxied URL starts with `http://localhost/site-utl` instead of `http://192.168.8.8/site-ur` ones.

I propose to use the common phrase: `To fix the problem, launch TestCafe with...`

Reviewer 3: > To use `localhost` in TestCafe URLs, ...

?

Reviewer 2: For which purposes I should `To use localhost in TestCafe URLs, ..`?

My variant tells why you should use it.

Reviewer 3: You have a point

Author: ```suggestion
To fix the issue, launch TestCafe with the `--hostname localhost` CLI option:
```

* [TravisCI](../documentation/guides/continuous-integration/travis.md)
* [CircleCI](../documentation/guides/continuous-integration/circleci.md)
* [AppVeyor](../documentation/guides/continuous-integration/appveyor.md)

### 'The browser can't open the page: can't establish a secure connection to the server'

If your computer is connected to multiple networks (for instance, uses a VPN connection), TestCafe may host its internal proxy server on an IP that cannot be routed correctly due to VPN. In this case, the following error occurs:

```text
<browsername> can't open the page because <browsername> can't establish a secure connection to the server.
```

> 

Author: ```suggestion
```

* [TravisCI](../documentation/guides/continuous-integration/travis.md)
* [CircleCI](../documentation/guides/continuous-integration/circleci.md)
* [AppVeyor](../documentation/guides/continuous-integration/appveyor.md)

### 'The browser can't open the page: can't establish a secure connection to the server'

> If the tested website is served via VPN, TestCafe may be unable to detect an IP address from where to retrieve the website. In this case, the following error occurs:

Author: ```suggestion
If your computer is connected to several networks (for instance, uses a VPN connection), TestCafe may host its internal proxy server on an IP that cannot be routed correctly due to VPN. In this case, the following error occurs:
```

```text
<browsername> can't open the page because <browsername> can't establish a secure connection to the server.
```

> This can happen when the machine is connected to multiple networks (for example, if you use a VPN).

Author: ```suggestion

```

> In this case TestCafe may not be able to detect an IP address that should be used to serve a tested website.

Author: ```suggestion

```

> To explicitly state that the tested website should be served on the local machine, launch TestCafe with the `--hostname localhost` CLI option:

Author: ```suggestion
To explicitly state that the proxy should be hosted on the local machine, launch TestCafe with the `--hostname localhost` CLI option:
```

* [TravisCI](../documentation/guides/continuous-integration/travis.md)
* [CircleCI](../documentation/guides/continuous-integration/circleci.md)
* [AppVeyor](../documentation/guides/continuous-integration/appveyor.md)

### 'Browser can't open the page: can't establish a secure connection to the server'

> If the tested website is served via OpenVPN, TestCafe may be unable to detect an IP address from where to retrieve the website. In this case, the following error occurs:

Reviewer 3: > OpenVPN

I thought OpenVPN is a brand, isn't it? Is this issue product-specific?

Author: You're right, the issue doesn't seem to be product specific.

```text

> <browsername> can't open the page "https://localhost:<port>/browser/connect/<windowid>"

Reviewer 3: > https://localhost

Why is `localhost` in this message?
I thought that TestCafe tries to open `localhost` only when you specify `--hostname localhost` explicitly (an approach that you propose as a solution). Otherwise, it uses an IP.

Is this behavior specific to the VPN issue (that TestCafe uses `localhost` by default)?

Author: > Otherwise, it uses an IP.

Seems like sometimes TestCafe picks an IP that belongs to the VPN connection (which leads to the aforementioned error).

because <browsername> can't establish a secure connection to the server "localhost"
```

This can happen when the machine is connected to multiple networks (for example, if you use a VPN).
In this case TestCafe may not be able to detect an IP address that should be used to serve a tested website.

To explicitly state that the tested website should be served on the local machine, launch TestCafe with the `--hostname localhost` CLI option:

```sh

> testcafe "chrome --hostname localhost" test.js

Reviewer 3: Are you intended to launch TestCafe with the `--hostname localhost` flag? Because this code tries to launch Chrome with `--hostname localhost`, doesn't it?

Author: You're right!
